pools/usecase/pools_usecase.go

```diff
@@ -540,7 +546,78 @@ func (p *poolsUseCase) CalcExitCFMMPool(poolID uint64, exitingShares osmomath.In
 
 	// fine to pass empty context as no data is mutated
 	exitFee := pool.GetExitFee(sdk.Context{})
-	return pool.CalcExitPoolCoinsFromShares(sdk.Context{}, exitingShares, exitFee)
+
+	return calcExitPool(sdk.Context{}, pool, exitingSharesIn, exitFee)
+}
+
+// errMsgFormatSharesLargerThanMax is the error message format for when the exiting shares are larger than the max allowed.
+const errMsgFormatSharesLargerThanMax = "cannot exit all shares in a pool. Attempted to exit %f shares, max allowed is %f"
+
+// calcExitPool is a helper function to calculate the exit pool.
+// It is a direct port of the CalcExitPool function from the Gamm module with some modifications
+// to optimize the calculation for performance.
+func calcExitPool(ctx sdk.Context, pool types.CFMMPoolI, exitingSharesIn osmomath.Int, exitFee osmomath.Dec) (sdk.Coins, error) {
+	totalShares, err := strconv.ParseFloat(pool.GetTotalShares().String(), 64)
+	if err != nil {
+		return sdk.Coins{}, err
+	}
+
+	exitingShares, err := strconv.ParseFloat(exitingSharesIn.String(), 64)
+	if err != nil {
+		return sdk.Coins{}, err
+	}
+
+	if exitingShares >= totalShares {
+		return sdk.Coins{}, errorsmod.Wrapf(types.ErrLimitMaxAmount, errMsgFormatSharesLargerThanMax, exitingShares, totalShares-float64(osmomath.OneInt().Int64()))
+	}
+
+	// refundedShares = exitingShares * (1 - exit fee)
+	// with 0 exit fee optimization
+	var refundedShares float64
+	if !exitFee.IsZero() {
+		f, err := exitFee.Float64()
+		if err != nil {
+			return sdk.Coins{}, err
+		}
+
+		// exitingShares * (1 - exit fee)
+		oneSubExitFee := 1.0 - f
+		refundedShares = exitingShares * oneSubExitFee
+	} else {
+		refundedShares = exitingShares
+	}
+
+	shareOutRatio := refundedShares / totalShares
+
+	// exitedCoins = shareOutRatio * pool liquidity
+	poolLiquidity := pool.GetTotalPoolLiquidity(ctx)
+	exitedCoins := make(sdk.Coins, 0, len(poolLiquidity))
+
+	for _, asset := range poolLiquidity {
+		// round down here, due to not wanting to over-exit
+		amount := float64(asset.Amount.BigInt().Int64())
+
+		exitAmt := shareOutRatio * amount
+		if exitAmt <= 0 {
+			continue
+		}
+
+		if exitAmt >= amount {
+			return sdk.Coins{}, errors.New("too many shares out")
+		}
+
+		// If the exit amount is within 1e-9 of an integer, round it to the nearest integer
+		if sdkmath.Abs(exitAmt-sdkmath.Round(exitAmt)) < 1e-9 {
+			exitAmt = (sdkmath.Round(exitAmt))
+		}
```

Comment:
Should we truncate in the pool's favor in case of rounding to the nearest int?

The general rule is that value out should be rounded down. Value in should be rounded up

Reply:
I was aiming to make original test case`TestCalcExitPool` to pass after after conversion to floats, to ensure there is no errors in the conversion logic and `TestCalcExitPool` functions as it was before making current changes.

There are two cases with rounding up/down, for example:

```
// up
89.99999999999999 - should become to 90
119.99999999999999 - should become to 120
59.99999999999999  - should become to 60

// and

// down
209.958 - should become to 209
279.944 - should become to 279
139.972 - should become to 139
```

Omitting rounding logic above tests would start complain as follows;

```
--- FAIL: TestPoolsUsecaseTestSuite (1.37s)
    --- FAIL: TestPoolsUsecaseTestSuite/TestCalcExitPool (0.01s)
        pools_usecase_test.go:707: 
                Error Trace:    /home/deividas/go/src/github.com/deividaspetraitis/sqs/pools/usecase/pools_usecase_test.go:707
                Error:          Not equal: 
                                expected: "900000bar,1200000baz,600000foo"
                                actual  : "899999bar,1200000baz,600000foo"
```

Reply:
I see - I'm happy to keep this as is since this is only needed for display so rounding behavior here is not critical.

Thanks for looking into it!